File: FitProj/app/src/main/java/com/teuskim/fitproj/MainDataFragment.java
package com.teuskim.fitproj;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.inputmethod.InputMethodManager;
import android.widget.BaseAdapter;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.Switch;
import android.widget.TextView;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.fitness.Fitness;
import com.google.android.gms.fitness.data.Bucket;
import com.google.android.gms.fitness.data.DataPoint;
import com.google.android.gms.fitness.data.DataSet;
import com.google.android.gms.fitness.data.DataSource;
import com.google.android.gms.fitness.data.DataType;
import com.google.android.gms.fitness.data.Field;
import com.google.android.gms.fitness.request.DataReadRequest;
import com.google.android.gms.fitness.result.DataReadResponse;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.teuskim.fitproj.common.FitPreference;
import com.teuskim.fitproj.common.FitUtil;
import com.teuskim.fitproj.view.FitViewPager;
import com.teuskim.fitproj.view.HorizontalListView;
import com.teuskim.fitproj.view.WeightItemView;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.concurrent.TimeUnit;


/**
 * 메인탭의 데이터 프레그먼트
 * 체중 등 사용자 정보 입력
 */
public class MainDataFragment extends BaseFragment {

    private static final String TAG = "WalkHolic";
    private EditText lastWeightView;
    private TextView lastWeightUnitView;
    private View btnModifyWrapper;
    private TextView lastDateView;
    private HorizontalListView weightListView;
    private View enterWeightMsgView;

    private WeightAdapter adapter;
    private Weight lastWeight;
    private FitPreference pref;
    private boolean isClickedWeight;
    private boolean isConvertUnitOn;

    public static MainDataFragment newInstance() {
        return new MainDataFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        pref = FitPreference.getInstance(getActivity());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        isClickedWeight = false;
        isConvertUnitOn = pref.isConvertUnitOn();
        View v = inflater.inflate(R.layout.main_data_fragment, container, false);
        initViews(v);
        loadData();
        ((MainActivity)getActivity()).setViewPagerOnChildMoveListener(new FitViewPager.OnChildMoveListener() {
            @Override
            public boolean onMove(float diffX) {
                int x = weightListView.getCurrentX();
                return !(x == 0 && diffX > 0);
            }
        });
        return v;
    }

    private void initViews(View v) {
        lastWeightView = v.findViewById(R.id.last_weight);
        lastWeightView.addTextChangedListener(new TextWatcher() {

            @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                if (!isClickedWeight) {
                    return;
                }
                String wstr = s.toString().trim();
                try {
                    float w = Float.parseFloat(wstr);
                    if ((lastWeight == null || (w > 0 && w != lastWeight.weight))
                            && btnModifyWrapper.getVisibility() != View.VISIBLE) {
                        Animation anim = AnimationUtils.loadAnimation(getActivity(), R.anim.right_in);
                        btnModifyWrapper.startAnimation(anim);
                        btnModifyWrapper.setVisibility(View.VISIBLE);
                    }
                } catch (Exception e) {
                    Log.e(TAG, "weight view text changed error", e);
                }
            }
        });
        lastWeightView.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (hasFocus) {
                    isClickedWeight = true;
                }
            }
        });

        lastWeightUnitView = v.findViewById(R.id.last_weight_unit);

        View.OnClickListener l = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                switch (v.getId()) {
                    case R.id.btn_modify:
                        InputMethodManager imm = (InputMethodManager)getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
                        if (imm != null) {
                            imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
                        }
                        saveWeight();
                        break;
                }
            }
        };
        btnModifyWrapper = v.findViewById(R.id.btn_modify_wrapper);
        v.findViewById(R.id.btn_modify).setOnClickListener(l);
        lastDateView = v.findViewById(R.id.last_date);
        weightListView = v.findViewById(R.id.weight_list);
        enterWeightMsgView = v.findViewById(R.id.enter_weight_img);
        adapter = new WeightAdapter();
        weightListView.setAdapter(adapter);

        Switch switchRecordingView = v.findViewById(R.id.switch_recording);
        switchRecordingView.setChecked(pref.isRecordingOn());
        switchRecordingView.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                pref.setRecordingOn(isChecked);
                ((MainActivity)getActivity()).subscribeOrCancel();
            }
        });

        Switch switchAlertView = v.findViewById(R.id.switch_alert);
        switchAlertView.setChecked(pref.isAlarmOn());
        switchAlertView.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                pref.setAlarmOn(isChecked);
                ((MainActivity)getActivity()).initAlarm();
            }
        });

        Switch switchConvertUnitView = v.findViewById(R.id.switch_convert_unit);
        switchConvertUnitView.setChecked(pref.isConvertUnitOn());
        switchConvertUnitView.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                pref.setConvertUnitOn(isChecked);
                isConvertUnitOn = isChecked;
                loadData();
            }
        });
    }

    private void loadData() {
        if (isConvertUnitOn) {
            lastWeightUnitView.setText("lbs");
        } else {
            lastWeightUnitView.setText("kg");
        }

        // 요청 만들기
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        long endTime = cal.getTimeInMillis() + 86400000L;
        long startTime = endTime - (86400000L * 365);
        DataReadRequest req = new DataReadRequest.Builder()
                .aggregate(DataType.TYPE_WEIGHT, DataType.AGGREGATE_WEIGHT_SUMMARY)
                .bucketByTime(1, TimeUnit.DAYS)
                .setTimeRange(startTime, endTime, TimeUnit.MILLISECONDS)
                .build();

        // 데이터 가져오기
        Fitness.getHistoryClient(getContext(), GoogleSignIn.getLastSignedInAccount(getContext()))
                .readData(req)
                .addOnSuccessListener(
                        new OnSuccessListener<DataReadResponse>() {
                            @SuppressLint("DefaultLocale")
                            @Override
                            public void onSuccess(DataReadResponse dataReadResponse) {
                                List<Weight> weightList = new ArrayList<>();
                                @SuppressLint("SimpleDateFormat") SimpleDateFormat dateFormat = new SimpleDateFormat("MM.dd");
                                for (Bucket bk : dataReadResponse.getBuckets()) {
                                    List<DataSet> list = bk.getDataSets();
                                    for (DataSet ds : list) {
                                        for (DataPoint dp : ds.getDataPoints()) {
                                            if (dp.getDataType().getFields().contains(Field.FIELD_AVERAGE)) {
                                                long date = dp.getStartTime(TimeUnit.MILLISECONDS);
                                                String dateText = dateFormat.format(date);
                                                float weight = getConvertedWeight(dp.getValue(Field.FIELD_AVERAGE).asFloat());
                                                weightList.add(new Weight(weight, date, dateText));
                                            }
                                        }
                                    }
                                }
                                List<Weight> reverseWeightList = new ArrayList<>();
                                for (int i=weightList.size()-1; i>=0; i--) {
                                    reverseWeightList.add(weightList.get(i));
                                }
                                if (reverseWeightList.size() > 0) {
                                    Calendar cal = Calendar.getInstance();
                                    lastWeight = reverseWeightList.get(0);
                                    lastWeightView.setText(String.format("%.1f", lastWeight.weight));
                                    cal.setTimeInMillis(lastWeight.date);
                                    lastDateView.setText(DateFormat.getDateInstance(DateFormat.FULL).format(cal.getTime()));
                                    enterWeightMsgView.setVisibility(View.GONE);

                                    adapter.setList(reverseWeightList);
                                    adapter.notifyDataSetChanged();
                                } else {  // 몸무게 데이터가 없을때
                                    lastWeightView.setText("0.0");
                                    lastDateView.setText(R.string.enter_weight);
                                    enterWeightMsgView.setVisibility(View.VISIBLE);
                                }
                            }
                        })
                .addOnFailureListener(
                        new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                // There was a problem reading the data.
                            }
                        });
    }

    private float getConvertedWeight(float weightKg) {
        if (isConvertUnitOn) {
            return (float)(weightKg * 2.20462262);  // convert from kg to lbs
        }
        return weightKg;
    }

    private void saveWeight() {
        float w = 0;
        try {
            w = Float.parseFloat(lastWeightView.getText().toString().trim());
        } catch (Exception e) {
            Log.e(TAG, "save weight error", e);
        }

        if (w <= 0) {
            showToast(R.string.wrong_number);
            return;
        }

        if (isConvertUnitOn) {
            w = (float)(w * 0.45359237);  // convert from lbs to kg
        }

        DataSource dataSource = new DataSource.Builder()
                .setAppPackageName(getActivity())
                .setDataType(DataType.TYPE_WEIGHT)
                .setName("fitproj-weight")
                .setType(DataSource.TYPE_RAW)
                .build();

        DataSet dataSet = DataSet.create(dataSource);

        Calendar cal = Calendar.getInstance();
        cal.set(cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH), 0, 0, 0);
        long startTime = cal.getTimeInMillis();
        long endTime = startTime + 86400000L - 3600000L; // 0시에서 23시까지

        DataPoint dataPoint = dataSet.createDataPoint()
                .setTimeInterval(startTime, endTime, TimeUnit.MILLISECONDS);
        dataPoint.getValue(Field.FIELD_WEIGHT).setFloat(w);
        dataSet.add(dataPoint);

        Fitness.getHistoryClient(getContext(), GoogleSignIn.getLastSignedInAccount(getContext()))
                .insertData(dataSet)
                .addOnCompleteListener(
                        new OnCompleteListener<Void>() {
                            @Override
                            public void onComplete(@NonNull Task<Void> task) {
                                if (task.isSuccessful()) {
                                    showToast(R.string.save_success);
                                    btnModifyWrapper.setVisibility(View.GONE);
                                    loadData();
                                } else {
                                    // There was a problem inserting the dataset.
                                    showToast(R.string.common_error);
                                    Log.e(TAG, "save weight error", task.getException());
                                }
                            }
                        });
    }


    class Weight {
        long date;
        String dateText;
        float weight;

        Weight(float weight, long date, String dateText) {
            this.date = date;
            this.dateText = dateText;
            this.weight = weight;
        }
    }

    class WeightAdapter extends BaseAdapter {

        private List<Weight> list = new ArrayList<>();
        private float maxWeight, minWeight;
        private int itemWidth;

        public void setList(List<Weight> list) {
            this.list = list;
            maxWeight = 0;
            minWeight = 1000;
            for (Weight w : list) {
                if (w.weight > maxWeight) {
                    maxWeight = w.weight;
                }
                if (w.weight < minWeight) {
                    minWeight = w.weight;
                }
            }
            if (list.size() < 6) {
                itemWidth = FitUtil.getScreenWidth(getActivity()) / list.size();
            } else {
                itemWidth = FitUtil.getScreenWidth(getActivity()) / 6;
            }
        }

        @Override
        public int getCount() {
            return list.size();
        }

        @Override
        public Weight getItem(int position) {
            return list.get(position);
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            WeightItemView v;
            if (convertView == null) {
                v = new WeightItemView(getActivity());
                HorizontalListView.LayoutParams lp = new HorizontalListView.LayoutParams(itemWidth, HorizontalListView.LayoutParams.MATCH_PARENT);
                v.setLayoutParams(lp);
                v.setRange(maxWeight, minWeight);
            } else {
                v = (WeightItemView)convertView;
            }

            Weight w = getItem(position);
            v.setDateText(w.dateText);
            v.setWeight(w.weight, isConvertUnitOn);
            if (position > 0) {
                Weight prev = getItem(position-1);
                v.setPrevWeight(prev.weight);
            } else {
                v.setPrevWeight(0);
            }
            if (position < getCount()-1) {
                Weight next = getItem(position+1);
                v.setNextWeight(next.weight);
            } else {
                v.setNextWeight(0);
            }

            return v;
        }
    }

}

File: FitProj_kotlin/app/src/main/java/com/teuskim/fitproj/MainDataFragment.kt
package com.teuskim.fitproj

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.view.inputmethod.InputMethodManager
import android.widget.BaseAdapter
import android.widget.EditText
import android.widget.Switch
import android.widget.TextView
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.fitness.Fitness
import com.google.android.gms.fitness.data.DataSet
import com.google.android.gms.fitness.data.DataSource
import com.google.android.gms.fitness.data.DataType
import com.google.android.gms.fitness.data.Field
import com.google.android.gms.fitness.request.DataReadRequest
import com.teuskim.fitproj.common.FitPreference
import com.teuskim.fitproj.common.FitUtil
import com.teuskim.fitproj.view.FitViewPager
import com.teuskim.fitproj.view.HorizontalListView
import com.teuskim.fitproj.view.WeightItemView
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit


/**
 * 메인탭의 데이터 프레그먼트
 * 체중 등 사용자 정보 입력
 */
class MainDataFragment : BaseFragment() {
    private var lastWeightView: EditText? = null
    private var lastWeightUnitView: TextView? = null
    private var btnModifyWrapper: View? = null
    private var lastDateView: TextView? = null
    private var weightListView: HorizontalListView? = null
    private var enterWeightMsgView: View? = null

    private var adapter: WeightAdapter? = null
    private var lastWeight: Weight? = null
    private var pref: FitPreference? = null
    private var isClickedWeight: Boolean = false
    private var isConvertUnitOn: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        pref = FitPreference.getInstance(activity)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        isClickedWeight = false
        isConvertUnitOn = pref!!.isConvertUnitOn
        val v = inflater.inflate(R.layout.main_data_fragment, container, false)
        initViews(v)
        loadData()
        (activity as MainActivity).setViewPagerOnChildMoveListener(object: FitViewPager.OnChildMoveListener {
            override fun onMove(diffX: Float): Boolean {
                val x = weightListView!!.currentX
                return !(x == 0 && diffX > 0)
            }
        })
        return v
    }

    private fun initViews(v: View) {
        lastWeightView = v.findViewById(R.id.last_weight)
        lastWeightView!!.addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                if (!isClickedWeight) {
                    return
                }
                val wstr = s.toString().trim { it <= ' ' }
                try {
                    val w = java.lang.Float.parseFloat(wstr)
                    if ((lastWeight == null || w > 0 && w != lastWeight!!.weight) && btnModifyWrapper!!.visibility != View.VISIBLE) {
                        val anim = AnimationUtils.loadAnimation(activity, R.anim.right_in)
                        btnModifyWrapper!!.startAnimation(anim)
                        btnModifyWrapper!!.visibility = View.VISIBLE
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "weight view text changed error", e)
                }

            }
        })
        lastWeightView!!.onFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
            if (hasFocus) {
                isClickedWeight = true
            }
        }

        lastWeightUnitView = v.findViewById(R.id.last_weight_unit)

        val l = View.OnClickListener { v ->
            when (v.id) {
                R.id.btn_modify -> {
                    val imm = activity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                    imm?.hideSoftInputFromWindow(v.windowToken, 0)
                    saveWeight()
                }
            }
        }
        btnModifyWrapper = v.findViewById(R.id.btn_modify_wrapper)
        v.findViewById<View>(R.id.btn_modify).setOnClickListener(l)
        lastDateView = v.findViewById(R.id.last_date)
        weightListView = v.findViewById(R.id.weight_list)
        enterWeightMsgView = v.findViewById(R.id.enter_weight_img)
        adapter = WeightAdapter()
        weightListView!!.adapter = adapter

        val switchRecordingView = v.findViewById<Switch>(R.id.switch_recording)
        switchRecordingView.isChecked = pref!!.isRecordingOn
        switchRecordingView.setOnCheckedChangeListener { buttonView, isChecked ->
            pref!!.setRecordingOn(isChecked)
            (activity as MainActivity).subscribeOrCancel()
        }

        val switchAlertView = v.findViewById<Switch>(R.id.switch_alert)
        switchAlertView.isChecked = pref!!.isAlarmOn
        switchAlertView.setOnCheckedChangeListener { buttonView, isChecked ->
            pref!!.setAlarmOn(isChecked)
            (activity as MainActivity).initAlarm()
        }

        val switchConvertUnitView = v.findViewById<Switch>(R.id.switch_convert_unit)
        switchConvertUnitView.isChecked = pref!!.isConvertUnitOn
        switchConvertUnitView.setOnCheckedChangeListener { buttonView, isChecked ->
            pref!!.setConvertUnitOn(isChecked)
            isConvertUnitOn = isChecked
            loadData()
        }
    }

    private fun loadData() {
        if (isConvertUnitOn) {
            lastWeightUnitView!!.text = "lbs"
        } else {
            lastWeightUnitView!!.text = "kg"
        }

        // 요청 만들기
        val cal = Calendar.getInstance()
        cal.set(Calendar.HOUR_OF_DAY, 0)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        val endTime = cal.timeInMillis + 86400000L
        val startTime = endTime - 86400000L * 365
        val req = DataReadRequest.Builder()
                .aggregate(DataType.TYPE_WEIGHT, DataType.AGGREGATE_WEIGHT_SUMMARY)
                .bucketByTime(1, TimeUnit.DAYS)
                .setTimeRange(startTime, endTime, TimeUnit.MILLISECONDS)
                .build()

        // 데이터 가져오기
        Fitness.getHistoryClient(context, GoogleSignIn.getLastSignedInAccount(context))
                .readData(req)
                .addOnSuccessListener { dataReadResponse ->
                    val weightList = ArrayList<Weight>()
                    @SuppressLint("SimpleDateFormat") val dateFormat = SimpleDateFormat("MM.dd")
                    for (bk in dataReadResponse.buckets) {
                        val list = bk.dataSets
                        for (ds in list) {
                            for (dp in ds.dataPoints) {
                                if (dp.dataType.fields.contains(Field.FIELD_AVERAGE)) {
                                    val date = dp.getStartTime(TimeUnit.MILLISECONDS)
                                    val dateText = dateFormat.format(date)
                                    val weight = getConvertedWeight(dp.getValue(Field.FIELD_AVERAGE).asFloat())
                                    weightList.add(Weight(weight, date, dateText))
                                }
                            }
                        }
                    }
                    val reverseWeightList = ArrayList<Weight>()
                    for (i in weightList.indices.reversed()) {
                        reverseWeightList.add(weightList[i])
                    }
                    if (reverseWeightList.size > 0) {
                        val cal = Calendar.getInstance()
                        lastWeight = reverseWeightList[0]
                        lastWeightView!!.setText(String.format("%.1f", lastWeight!!.weight))
                        cal.timeInMillis = lastWeight!!.date
                        lastDateView!!.text = DateFormat.getDateInstance(DateFormat.FULL).format(cal.time)
                        enterWeightMsgView!!.visibility = View.GONE

                        adapter!!.setList(reverseWeightList)
                        adapter!!.notifyDataSetChanged()
                    } else {  // 몸무게 데이터가 없을때
                        lastWeightView!!.setText("0.0")
                        lastDateView!!.setText(R.string.enter_weight)
                        enterWeightMsgView!!.visibility = View.VISIBLE
                    }
                }
                .addOnFailureListener {
                    // There was a problem reading the data.
                }
    }

    private fun getConvertedWeight(weightKg: Float): Float {
        return if (isConvertUnitOn) {
            (weightKg * 2.20462262).toFloat()  // convert from kg to lbs
        } else weightKg
    }

    private fun saveWeight() {
        var w = 0f
        try {
            w = java.lang.Float.parseFloat(lastWeightView!!.text.toString().trim { it <= ' ' })
        } catch (e: Exception) {
            Log.e(TAG, "save weight error", e)
        }

        if (w <= 0) {
            showToast(R.string.wrong_number)
            return
        }

        if (isConvertUnitOn) {
            w = (w * 0.45359237).toFloat()  // convert from lbs to kg
        }

        val dataSource = DataSource.Builder()
                .setAppPackageName(activity)
                .setDataType(DataType.TYPE_WEIGHT)
                .setName("fitproj-weight")
                .setType(DataSource.TYPE_RAW)
                .build()

        val dataSet = DataSet.create(dataSource)

        val cal = Calendar.getInstance()
        cal.set(cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH), 0, 0, 0)
        val startTime = cal.timeInMillis
        val endTime = startTime + 86400000L - 3600000L // 0시에서 23시까지

        val dataPoint = dataSet.createDataPoint()
                .setTimeInterval(startTime, endTime, TimeUnit.MILLISECONDS)
        dataPoint.getValue(Field.FIELD_WEIGHT).setFloat(w)
        dataSet.add(dataPoint)

        Fitness.getHistoryClient(context, GoogleSignIn.getLastSignedInAccount(context))
                .insertData(dataSet)
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        showToast(R.string.save_success)
                        btnModifyWrapper!!.visibility = View.GONE
                        loadData()
                    } else {
                        // There was a problem inserting the dataset.
                        showToast(R.string.common_error)
                        Log.e(TAG, "save weight error", task.exception)
                    }
                }
    }


    internal inner class Weight(var weight: Float, var date: Long, var dateText: String)

    internal inner class WeightAdapter : BaseAdapter() {

        private var list: List<Weight> = ArrayList()
        private var maxWeight: Float = 0.toFloat()
        private var minWeight: Float = 0.toFloat()
        private var itemWidth: Int = 0

        fun setList(list: List<Weight>) {
            this.list = list
            maxWeight = 0f
            minWeight = 1000f
            for (w in list) {
                if (w.weight > maxWeight) {
                    maxWeight = w.weight
                }
                if (w.weight < minWeight) {
                    minWeight = w.weight
                }
            }
            if (list.size < 6) {
                itemWidth = FitUtil.getScreenWidth(activity) / list.size
            } else {
                itemWidth = FitUtil.getScreenWidth(activity) / 6
            }
        }

        override fun getCount(): Int {
            return list.size
        }

        override fun getItem(position: Int): Weight {
            return list[position]
        }

        override fun getItemId(position: Int): Long {
            return 0
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val v: WeightItemView
            if (convertView == null) {
                v = WeightItemView(activity)
                val lp = ViewGroup.LayoutParams(itemWidth, ViewGroup.LayoutParams.MATCH_PARENT)
                v.layoutParams = lp
                v.setRange(maxWeight, minWeight)
            } else {
                v = (convertView as WeightItemView?)!!
            }

            val w = getItem(position)
            v.setDateText(w.dateText)
            v.setWeight(w.weight, isConvertUnitOn)
            if (position > 0) {
                val prev = getItem(position - 1)
                v.setPrevWeight(prev.weight)
            } else {
                v.setPrevWeight(0f)
            }
            if (position < count - 1) {
                val next = getItem(position + 1)
                v.setNextWeight(next.weight)
            } else {
                v.setNextWeight(0f)
            }

            return v
        }
    }

    companion object {

        private val TAG = "WalkHolic"

        fun newInstance(): MainDataFragment {
            return MainDataFragment()
        }
    }

}
